### PR TITLE
Batch & Sparse Attention

### PR DESCRIPTION
**Pull request by Starmys**
- Refactor kernels & TeSA
- Add batch dimension
- Sparse attention function & operator

### CODE REVIEW

**Comment by Starmys** [comment]
> @Starmys please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree [company="microsoft"]

**Comment by Starmys** [comment]
@microsoft-github-policy-service agree

**Comment by Starmys** [comment]
@microsoft-github-policy-service agree company="microsoft"

**Comment by Starmys** [comment]
> @Starmys please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="microsoft"

**Comment by Starmys** [comment]
Test

**Comment by Starmys** [comment]
Test Comment

File: benchmark/compare_latency.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import unittest
```
[the remark this answers is not in the document]
Comment (Starmys) [reply benchmark/compare_latency.py]: Header added to all source codes.

File: sparta/common/tesa.py [hunk 2]
```diff
@@ -2,226 +2,178 @@
...
+    def get_mask(self) -> torch.Tensor:
+        '''Get the mask actually used.'''
+
+    def __call__(self, dense: torch.Tensor) -> torch.Tensor:
+        return self.convert(dense)
+
+
+class BCSR(TeSAConverter):
+
+    def __init__(self, mask: torch.Tensor, size: Tuple[int, int], block_size: Tuple[int, int]):
```
[the remark this answers is not in the document]
Comment (Starmys) [reply sparta/common/tesa.py]: [The new TeSA constructor](https://github.com/microsoft/SparTA/pull/13/commits/b3266c4848364366151e7f2f36fc0c222b53f7c3#diff-d349f4ce11d5fa1b786bbf240f0ee6d665450250cab90acc3a2eccb9e74ddd82)

```python
bcsr = BCSR(mask, block_size, device)
```

File: sparta/common/tesa.py [hunk 3]
```diff
@@ -2,226 +2,178 @@
...
-        return val, ptr, idx
+            sum_val = torch.zeros(size=(batch_size, self._W), device=sparse_val.device)
+            for k, col in enumerate(self.get_attr('col_idx')):
+                segment_start = col * self._BW
+                segment_end = segment_start + self._BW
+                sum_val[:, segment_start:segment_end] += sparse_val[:, k, :]
+        return sum_val
+
+
+class BCSRH(BCSR):
```
[the remark this answers is not in the document]
Comment (Starmys) [reply sparta/common/tesa.py]: [BSCR Unit Test](https://github.com/microsoft/SparTA/pull/13/commits/b3266c4848364366151e7f2f36fc0c222b53f7c3#diff-ec25965f63ccdf3799a728afc9d7816220b5f06b4e75226e79147e44aa0b9d28)

File: sparta/common/tesa.py [hunk 4]
```diff
@@ -2,226 +2,178 @@
...
+class BCSR(TeSAConverter):
+
+    def __init__(self, mask: torch.Tensor, size: Tuple[int, int], block_size: Tuple[int, int]):
+        super().__init__()
+        self._H, self._W = size
+        self._BH, self._BW = block_size
+        self._block_size = self._BH * self._BW
+        row_num = self._H // self._BH
+        col_num = self._W // self._BW
+        self._block_mask = mask.reshape((row_num, self._BH, col_num, self._BW))
```
[the remark this answers is not in the document]
Comment (Starmys) [reply sparta/common/tesa.py:58]: [The new TeSA constructor](https://github.com/microsoft/SparTA/pull/13/commits/b3266c4848364366151e7f2f36fc0c222b53f7c3#diff-d349f4ce11d5fa1b786bbf240f0ee6d665450250cab90acc3a2eccb9e74ddd82)

```python
bcsr = BCSR(mask, block_size, device)
```

File: sparta/common/tesa.py [hunk 5]
```diff
@@ -2,226 +2,178 @@
...
+        self._block_mask = self._block_mask.swapaxes(1, 2).any(dim=-1).any(dim=-1)
+        attr_names = ('row_idx', 'col_idx', 'row_ptr', 'col_ptr', 'nnz')
+        for name, value in zip(attr_names, self.read_block_mask(self._block_mask)):
+            self._set_attr(name, torch.tensor(value, dtype=torch.int32, device='cuda'))
```
[the remark this answers is not in the document]
Comment (Starmys) [reply sparta/common/tesa.py]: [The new TeSA constructor](https://github.com/microsoft/SparTA/pull/13/commits/b3266c4848364366151e7f2f36fc0c222b53f7c3#diff-d349f4ce11d5fa1b786bbf240f0ee6d665450250cab90acc3a2eccb9e74ddd82)

```python
bcsr = BCSR(mask, block_size, device)
```

File: sparta/specializer/funtional/batch_matmul.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                    conditions[k][kernel_name] = f'BLOCK_SIZE_{dim}_VALUE'
+            return conditions
+        else:
+            return []
+
+    def dense_forward(self, *args):
+        return self._kernels['forward:C'].dense_func(*args)
+
+
+class SparseBatchMatMulFunc(torch.autograd.Function):
```
Comment (mydmdm) [inline sparta/specializer/funtional/batch_matmul.py:159]: please double check whether this function behaves like the sparse version of [torch.bmm](https://pytorch.org/docs/stable/generated/torch.bmm.html)
if yes, please give some docstring like 
> If `input` is a (b×n×m) tensor, `mat2` is a (b×m×p) tensor, `out` will be a (b×n×p) tensor.
> out_i = input_i @ mat2_i

**Review by mydmdm — APPROVED** [review]

**Pull request merged**